### ISSUE TITLE
Hello, why does the sample preview in aitoolkit look great, but the same prompt words look much worse in ComfyUI? How can I reproduce this in ComfyUI? Thank you!

### ISSUE THREAD

**Issue by icefox21** [ostris/ai-toolkit#589]
I can't achieve the same effect in ComfyUI as in the AItoolkit preview, no matter what I do. I'm using a trained zimage turbo lora.

**Comment by icefox21** [comment]
In the sampling preview of AIToolkit, both prompt adherence and detail adherence are extremely flawless—everything that should be present is there, and everything that shouldn’t be is completely absent. But no matter how I adjust the settings in ComfyUI, the result is far from satisfactory. It’s really perplexing. Could you please help me? Thank you. I’ve also met friends in the community who are facing the same issue, and we’re all confused.

**Comment by FinallyKiKi** [comment]
Same problem in Zimage training. 
The sample in aitoolkit looks better, but when I use the same (prompt, step, guidance_scale and seed) in my code, it gives a different image.

This might be related to the quantization operation I performed. The original Zimage pipe uses no more than 24GB of VRAM, allowing me to load it directly without quantization. However, when loading via aitoolkit, 7-bit quantization becomes mandatory.

The sample code:
```
pipe = ZImagePipeline.from_pretrained(
    "/my/path/to/Z-Image-Turbo",
    torch_dtype=torch.bfloat16,
    low_cpu_mem_usage=True
).to('cuda')
pipe.load_lora_weights("/my/path/to/Z-Image-Turbo",weight_name=/my/path/to/lora.safetensor)
image = pipe(
    prompt=prompt,
    negative_prompt=neg_prompt,
    height=1024,
    width=1024,
    num_inference_steps=20,  
    guidance_scale=3.0,
    generator=torch.Generator("cuda").manual_seed(42),
).images[0]
```
In the aitoolkit, the default parameters of guidance_scale=3 and step=20 deliver optimal performance, while in my code, following the Zimage author's recommended values (guidance_scale=0, step=24) yields the best results. This discrepancy is perplexing given that all other parameters remain identical between the two implementations.

Can the Zimage training code be optimized to use the pipe provided by the author? Or can a sample output method be provided to generate images identical to those produced during training?

**Comment by FinallyKiKi** [comment]
I've solved the problem. The root cause was that the image generator was resetting its seed with each use—it should instead be explicitly assigned a fixed seed every time it's utilized.
For example, the model Zimage trainning: 
`extensions_built_in/diffusion_models/z_image/z_image.py, line 309-319`
```
img = pipeline(
    prompt_embeds=conditional_embeds.text_embeds,
    negative_prompt_embeds=unconditional_embeds.text_embeds,
    height=gen_config.height,
    width=gen_config.width,
    num_inference_steps=gen_config.num_inference_steps,
    guidance_scale=gen_config.guidance_scale,
    latents=gen_config.latents,
    generator=generator.manual_seed(gen_config.seed),    # need reset seed
    **extra,
).images[0]
```
And if you give specific seed, will get corresponding image.

**Comment by dariiboi** [comment]
would also recommend trying the same scheduler as is used in AITK validation.
aitk uses flowmatch euler discrete, which is by default not available in comfy.
https://github.com/erosDiffusion/ComfyUI-EulerDiscreteScheduler this custom nodepack  should provide the scheduler in comfy

**Comment by FurkanGozukara** [comment]
> would also recommend trying the same scheduler as is used in AITK validation. aitk uses flowmatch euler discrete, which is by default not available in comfy. https://github.com/erosDiffusion/ComfyUI-EulerDiscreteScheduler this custom nodepack should provide the scheduler in comfy

i just tested this and this is worse than what i use

any other parameter needs to be set?

**Comment by KBerwari** [comment]
> would also recommend trying the same scheduler as is used in AITK validation. aitk uses flowmatch euler discrete, which is by default not available in comfy. https://github.com/erosDiffusion/ComfyUI-EulerDiscreteScheduler this custom nodepack should provide the scheduler in comfy

Might be wrong but i believe if you use Euler-Simple in comfy that is identical as it enables flow-match for z-image. This custom node is still useful for customizing as it exposes more controls.

**Comment by icefox21** [comment]
> > would also recommend trying the same scheduler as is used in AITK validation. aitk uses flowmatch euler discrete, which is by default not available in comfy. https://github.com/erosDiffusion/ComfyUI-EulerDiscreteScheduler this custom nodepack should provide the scheduler in comfy
> 
> Might be wrong but i believe if you use Euler-Simple in comfy that is identical as it enables flow-match for z-image. This custom node is still useful for customizing as it exposes more controls.

This problem has actually existed for at least a year. I've seen several people on Reddit raise this issue, but none of them have found a solution. I'm retraining it today, using an NSFW LoRa dataset. Again, from the first preview sample after 300 steps, the effect is excellent, because we all know Zimage doesn't characterize NSFW well. However, in the aitoolkit preview, it looks great. I've now trained it for 4200 steps, and I tried it in ComfyUI. The result is still unacceptable; the distortion is severe. I used the [EulerDiscreteScheduler](https://github.com/erosDiffusion/ComfyUI-EulerDiscreteScheduler) you mentioned a week ago. I was worried that bf16 might not be the same as the diffuser used by aitoolkit, so I wrote a node that can directly load the zimage diffuser folder. But even under these conditions, with the same seed, the effect is completely different. I'm completely stumped.

**Comment by icefox21** [comment]
> I've solved the problem. The root cause was that the image generator was resetting its seed with each use—it should instead be explicitly assigned a fixed seed every time it's utilized.我已经解决了这个问题。根本原因是图像生成器每次使用都会重置种子——它应该每次使用时都显式地分配一个固定的种子。 For example, the model Zimage trainning:例如，模型 Zimage 的训练： `extensions_built_in/diffusion_models/z_image/z_image.py, line 309-319`
> 
> ```
> img = pipeline(
>     prompt_embeds=conditional_embeds.text_embeds,
>     negative_prompt_embeds=unconditional_embeds.text_embeds,
>     height=gen_config.height,
>     width=gen_config.width,
>     num_inference_steps=gen_config.num_inference_steps,
>     guidance_scale=gen_config.guidance_scale,
>     latents=gen_config.latents,
>     generator=generator.manual_seed(gen_config.seed),    # need reset seed
>     **extra,
> ).images[0]
> ```
> 
> And if you give specific seed, will get corresponding image.如果提供特定的种子，就能得到对应的图像。

Hello, can you completely reproduce the image in the aitoolkit preview? I've tried everything, but nothing works. The seed you mentioned isn't the one I understand correctly; during training, the seed was 42. I've tried many different sampler debugger combinations with seed 42, but I still can't achieve the same effect as in the preview, and it's far from perfect.

**Comment by FinallyKiKi** [comment]
Yes, I can fully reproduce the samples generated by aitoolkit. What I mean is that aitoolkit has a bug—it runs with a random seed even when a fixed seed is set.
I've provided code that fixes this bug (for Zimage only); you can copy this code into the corresponding location in your code.

In my understanding, the parameters affecting image generation include these: size, step, scale, and seed. When using the same model, generating identical images only requires these parameters to match exactly.
You can regenerate a set of samples using aitoolkit (with the "force sample generation" option), then test with the same parameters in ComfyUI. For more precise variable control, however, I recommend generating directly via code rather than ComfyUI—this eliminates potential hidden variables introduced by GUI-based workflows, such as implicit preprocessing steps or version mismatches in underlying libraries.

If you encounter issues while following these steps, feel free to contact me.

**Comment by jremb** [comment]
The explanation makes little sense, or maybe I'm misunderstanding the claim being made. If there was a bug where it was using a random seed, then we should see completely different images between step 0 and step 250, step 500, etc. Clearly, the same seed is being used throughout training and between different training sessions, because I can see the same result between a v1 LoRA and a v2 LoRA where other parameters have changed. The images initial samples consistent, so it is the same seed.

There may be some confusion for some users who have `walk_seed: true` in their configuration, which means if the first sample prompt is generated with seed 42, then the second sample prompt is generated with seed 43, etc. Perhaps some don't realize this parameter is set, and thus are confused as to what seed they should be using in ComfyUI.

However, there is clearly a difference between the ai-toolkit output and the ComfyUI output. It's not clear if this is due to quantization or differences in diffusers etc.

**Comment by icefox21** [comment]
Yes, it is confusing and helpless, I tried to use AI analysis, and the result should be beyond its ability range.

---- Replied Message ----
| From | ***@***.***> |
| Date | 12/26/2025 07:45 |
| To | ostris/ai-toolkit ***@***.***> |
| Cc | icefox21 ***@***.***>,
Author ***@***.***> |
| Subject | Re: [ostris/ai-toolkit] Hello, why does the sample preview in aitoolkit look great, but the same prompt words look much worse in ComfyUI? How can I reproduce this in ComfyUI? Thank you! (Issue #589) |
jremb left a comment (ostris/ai-toolkit#589)

The explanation makes little sense, or maybe I'm misunderstanding the claim being made. If there was a bug where it was using a random seed, then we should see completely different images between step 0 and step 250, step 500, etc. Clearly, the same seed is being used throughout training and between different training sessions, because I can see the same result between a v1 LoRA and a v2 LoRA where other parameters have changed. The images initial samples consistent, so it is the same seed.

There may be some confusion for some users who have walk_seed: true in their configuration, which means if the first sample prompt is generated with seed 42, then the second sample prompt is generated with seed 43, etc. Perhaps some don't realize this parameter is set, and thus are confused as to what seed they should be using in ComfyUI.

However, there is clearly a difference between the ai-toolkit output and the ComfyUI output. It's not clear if this is due to quantization or differences in diffusers etc.

**Comment by FinallyKiKi** [comment]
This is a bug. The issue still occurs even when walk_seed=False. Cause seeds always change according to a certain pattern, resulting in almost identical sample images between v1 and v2. You might as well try modifying the code as I suggested first. 
I'll copy the code mentioned earlier.
`extensions_built_in/diffusion_models/z_image/z_image.py, line 309-319`
```
img = pipeline(
    prompt_embeds=conditional_embeds.text_embeds,
    negative_prompt_embeds=unconditional_embeds.text_embeds,
    height=gen_config.height,
    width=gen_config.width,
    num_inference_steps=gen_config.num_inference_steps,
    guidance_scale=gen_config.guidance_scale,
    latents=gen_config.latents,
    generator=generator.manual_seed(gen_config.seed),    # need reset seed
    **extra,
).images[0]
```

AI isn't omnipotent—it still can't solve such complex problems at present.

**Comment by icefox21** [comment]
> This is a bug. The issue still occurs even when walk_seed=False. Cause seeds always change according to a certain pattern, resulting in almost identical sample images between v1 and v2. You might as well try modifying the code as I suggested first. I'll copy the code mentioned earlier. `extensions_built_in/diffusion_models/z_image/z_image.py, line 309-319`
> 
> ```
> img = pipeline(
>     prompt_embeds=conditional_embeds.text_embeds,
>     negative_prompt_embeds=unconditional_embeds.text_embeds,
>     height=gen_config.height,
>     width=gen_config.width,
>     num_inference_steps=gen_config.num_inference_steps,
>     guidance_scale=gen_config.guidance_scale,
>     latents=gen_config.latents,
>     generator=generator.manual_seed(gen_config.seed),    # need reset seed
>     **extra,
> ).images[0]
> ```
> 
> AI isn't omnipotent—it still can't solve such complex problems at present.

Thank you for such a great answer. I really can't get the answer from AI. Also, let me explain the specific problem I'm encountering. I'm training a character named Lora, and the training set includes a certain number of NSFW images. I'm trying to correct Zimage's shortcomings in this area through image editing. In fact, I've seen very good preview images of genitals in AIToolkit, but in ComfyUI, after using many methods, the quality of the genitals in the resulting images is very poor. Actually, if it were just the character Lora, it wouldn't be so confusing for me. It just feels like AIToolkit has special methods for generating images, as if it bypasses the limitations of Zimage Turbo.

**Comment by FinallyKiKi** [comment]
Aitoolkit doesn't have specialized model processing and image generation mechanisms. 
Notably, its default parameters are set to (scale=3.0, step=20). In contrast, the official recommended values are (scale=0.0, step=9). In my LoRA, it is set to (scale=0.0, step=20). I suggest testing combinations in your specific use case.


> Thank you for such a great answer. I really can't get the answer from AI. Also, let me explain the specific problem I'm encountering. I'm training a character named Lora, and the training set includes a certain number of NSFW images. I'm trying to correct Zimage's shortcomings in this area through image editing. In fact, I've seen very good preview images of genitals in AIToolkit, but in ComfyUI, after using many methods, the quality of the genitals in the resulting images is very poor. Actually, if it were just the character Lora, it wouldn't be so confusing for me. It just feels like AIToolkit has special methods for generating images, as if it bypasses the limitations of Zimage Turbo.

**Comment by icefox21** [comment]
Thank you for your reply, I don't really understand why the official recommends cfg=0, step=9, but the official zimage workflow uses cfg=1 step=9, (you also use cfg=0) I have also tried to use cfg=0 and step=9 in comfyui but the generated image is very poor, can you help me clear my doubts, thank you. I have also asked AI this question, and as you did, it can't answer it. Also, I just tested the LORA generated in AITOOLKIT in comfyui with CFG=3 step=20, and it is indeed better than the previous use of CFG=1 and STEP=9. It's just that the color is a bit weird, but it's a big takeaway, I'm going to test it in depth, thank you again . Looking forward to your reply as well.



At 2025-12-26 10:43:31, "FinallyKiKi" ***@***.***> wrote:

FinallyKiKi left a comment (ostris/ai-toolkit#589)

Aitoolkit doesn't have specialized model processing and image generation mechanisms.
Notably, its default parameters are set to (scale=3.0, step=20). In contrast, the official recommended values are (scale=0.0, step=9). In my LoRA, it is set to (scale=0.0, step=20). I suggest testing combinations in your specific use case.

Thank you for such a great answer. I really can't get the answer from AI. Also, let me explain the specific problem I'm encountering. I'm training a character named Lora, and the training set includes a certain number of NSFW images. I'm trying to correct Zimage's shortcomings in this area through image editing. In fact, I've seen very good preview images of genitals in AIToolkit, but in ComfyUI, after using many methods, the quality of the genitals in the resulting images is very poor. Actually, if it were just the character Lora, it wouldn't be so confusing for me. It just feels like AIToolkit has special methods for generating images, as if it bypasses the limitations of Zimage Turbo.

**Comment by jremb** [comment]
> This is a bug. The issue still occurs even when walk_seed=False. Cause seeds always change according to a certain pattern, resulting in almost identical sample images between v1 and v2. You might as well try modifying the code as I suggested first. I'll copy the code mentioned earlier. `extensions_built_in/diffusion_models/z_image/z_image.py, line 309-319`
> 
> ```
> img = pipeline(
>     prompt_embeds=conditional_embeds.text_embeds,
>     negative_prompt_embeds=unconditional_embeds.text_embeds,
>     height=gen_config.height,
>     width=gen_config.width,
>     num_inference_steps=gen_config.num_inference_steps,
>     guidance_scale=gen_config.guidance_scale,
>     latents=gen_config.latents,
>     generator=generator.manual_seed(gen_config.seed),    # need reset seed
>     **extra,
> ).images[0]
> ```
> 
> AI isn't omnipotent—it still can't solve such complex problems at present.

No. I tried your suggestion and the sample images generated are identical to what they generated previously. Meaning, the seed was not different than what was set in the config.

**Comment by mert-cokelek** [comment]
removing `"cuda"` from `gen = torch.Generator("cuda").manual_seed(SEED)` solved the issue for me.

**Comment by icefox21** [comment]
Hello, could you please explain the specific steps? In aitoolkit or comfyui? Thank you. This seems difficult to understand; why would modifying this affect the preview or the final generated result?


> removing `"cuda"` from `gen = torch.Generator("cuda").manual_seed(SEED)` solved the issue for me.

**Comment by mert-cokelek** [comment]
I am using custom inference script, and I was able to reproduce the samples generated during training with this change, on top of @FinallyKiKi's proposal. Sorry, I did not try toolkit or comfyui.

**Comment by icefox21** [comment]
> I am using custom inference script, and I was able to reproduce the samples generated during training with this change, on top of [@FinallyKiKi](https://github.com/FinallyKiKi)'s proposal. Sorry, I did not try toolkit or comfyui.

Ok, thank you!
